// @vonage/applications/dist/applications.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Applications = void 0;
const tslib_1 = require("tslib");
const server_client_1 = require("@vonage/server-client");
class Applications extends server_client_1.Client {
    constructor() {
        super(...arguments);
        this.authType = server_client_1.AuthenticationType.BASIC;
    }
    createApplication(application) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resp = yield this.sendPostRequest(`${this.config.apiHost}/v2/applications`, application);
            return resp.data;
        });
    }
    deleteApplication(applicationId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.sendDeleteRequest(`${this.config.apiHost}/v2/applications/${applicationId}`);
        });
    }
    getApplication(applicationId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resp = yield this.sendGetRequest(`${this.config.apiHost}/v2/applications/${applicationId}`);
            return resp.data;
        });
    }
    listApplications(filter) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resp = yield this.sendGetRequest(`${this.config.apiHost}/v2/applications`, filter);
            return resp.data;
        });
    }
    updateApplication(application) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const resp = yield this.sendPutRequest(`${this.config.apiHost}/v2/applications/${application.id}`, application);
            return resp.data;
        });
    }
}
exports.Applications = Applications;
