// @vonage/accounts/dist/secrets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Secrets = void 0;
const tslib_1 = require("tslib");
const server_client_1 = require("@vonage/server-client");
class Secrets extends server_client_1.Client {
    constructor() {
        super(...arguments);
        this.authType = server_client_1.AuthenticationType.BASIC;
    }
    createSecret(apiKey, secret) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.sendPostRequest(`${this.config.apiHost}/accounts/${apiKey}/secrets`, { secret });
            return response.data;
        });
    }
    deleteSecret(apiKey, id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.sendDeleteRequest(`${this.config.apiHost}/accounts/${apiKey}/secrets/${id}`);
        });
    }
    getSecret(apiKey, id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.sendGetRequest(`${this.config.apiHost}/accounts/${apiKey}/secrets/${id}`);
            return response.data;
        });
    }
    listSecrets(apiKey) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.sendGetRequest(`${this.config.apiHost}/accounts/${apiKey}/secrets`);
            return response.data;
        });
    }
}
exports.Secrets = Secrets;
